### PR TITLE
Leaderboard website

### PR DESCRIPTION
**Pull request by noClaps**

### CODE REVIEW

**Comment by noClaps** [comment]
I've done the initial design with the test data so far, if you can review the design and let me know if you have any suggestions that would be great.

File: src/dashboard/script.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+let users: User[] = [];
+
+for (let i = 0; i < 10; i++) {
+  const page = await FetchPageTest(i);
+
+  if (page) {
+    users.push(...page);
+  }
+}
```
Comment (MesterMan03) [inline src/dashboard/script.ts:23]: This simply cannot be in production code, has to be replaced with infinite scrolling / a page selector.

File: src/dashboard/script.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    "beforeend",
+    `<div class="user">
+      <p class="pos">${user.pos}</p>
+      <p class="name" onclick="navigator.clipboard.writeText('${user.userid}')"><img src="${user.avatar}">${user.username}</p>
+      <label>
+        <progress value="${user.progress[0]}" max="${(user.progress[0] * 100) / user.progress[1]}">${Math.floor((user.progress[0] * 100) / user.progress[1])}</progress>
+        ${xpInfoHTML}
+      </label>
+    </div>`,
+  );
```
Comment (MesterMan03) [inline src/dashboard/script.ts:63]: I'd suggest adding a small popup indicating a successful copy of the User ID.

**Review by MesterMan03 — CHANGES_REQUESTED** [review]
Other stuff I found:
- The level of the user is completely missing.
- I'd put the total xp into the listing, not as a hidden part of the progress bar. This is an information most people would like to view the leaderboard for.
- No mobile support yet.
- Because of the random username lengths, the top 3 listings aren't always consistent. The long usernames push the other elements around, so I'd suggest adding a length cap (with an ellipsis) to prevent that from happening. Then #1 can be centered in the middle, while the #2 and #3 are chilling on the sides.

File: src/dashboard/script.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        <progress value="${user.progress[0]}" max="${(user.progress[0] * 100) / user.progress[1]}">${Math.floor((user.progress[0] * 100) / user.progress[1])}</progress>
+        ${xpInfoHTML}
+      </label>
+    </div>`,
+    );
+  }
+}
+
+let pageCursor = 0;
+const maxPage = 10;
```
Comment (MesterMan03) [inline src/dashboard/script.ts:63]: A maxPage will not work in production code. However, after removing it, the user should get some sort of error message when they scroll all the way down, for example "Sorry lad, you reached the end!"

File: src/dashboard/index.ts [hunk 4]
```diff
@@ -1,11 +1,49 @@
...
+
+        if (path === "/") return new Response(Bun.file(join(__dirname, indexLocation)));
+
+        if (path === "/script.js") {
+            return new Response(script, {
+                headers: { "Content-Type": "application/javascript" },
+            });
+        }
+
+        if (path === "/page") { // TODO: possible DOS without a ratelimit, add one?
```
Comment (MesterMan03) [inline src/dashboard/index.ts:30]: I'm not sure how to add ratelimiting in Bun, but it's very important.

File: src/dashboard/script.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+async function FetchPage(page: number): ReturnType<typeof IFetchPage> {
+  const res = await fetch(`/page?page=${page}`);
+  if (!res.ok) return null;
+
+  return res.json();
+}
+
+async function loadUsers(pageCursor: number) {
+  const page = await FetchPage(pageCursor);
+  if (!page) return;
```
Comment (MesterMan03) [inline src/dashboard/script.ts:14]: Loading with test data is incredibly quick, but when we reach production code, loading times might become noticeable. A small loading icon or text would be nice to let the user know the next page is being processed.

File: src/dashboard/script.ts (hunk 3 above)
Comment (noClaps) [reply src/dashboard/script.ts:63]: I was gonna replace that with the `GetMaxPage()` function you told me to use.

**Review by MesterMan03 — COMMENTED** [review]
New bugs:
- Scrolling up and down while a new page is being loaded will trigger the load again, causing the listings to appear multiple times.
- The "Loading..." text doesn't show up after the first page.

**Review by MesterMan03 — APPROVED** [review]
LGTM

**Pull request merged**